wolfcrypt/src/evp.c

```diff
@@ -1908,15 +1922,132 @@ int wolfSSL_EVP_PKEY_sign(WOLFSSL_EVP_PKEY_CTX *ctx, unsigned char *sig,
         }
 #endif /* NO_RSA */
 
-    case EVP_PKEY_EC:
-        WOLFSSL_MSG("not implemented");
-        FALL_THROUGH;
+#ifndef NO_DSA
+    case EVP_PKEY_DSA:
+        if (wolfSSL_DSA_do_sign(tbs, sig, ctx->pkey->dsa) == WOLFSSL_SUCCESS) {
+            *siglen = wolfSSL_BN_num_bytes(ctx->pkey->dsa->q);
+            return WOLFSSL_SUCCESS;
+        } else {
+            return WOLFSSL_FAILURE;
+        }
+#endif /* NO_DSA */
+
+#ifdef HAVE_ECC
+    case EVP_PKEY_EC: {
+        WOLFSSL_ECDSA_SIG *ecdsaSig = wolfSSL_ECDSA_do_sign(tbs, (int)tbslen, ctx->pkey->ecc);
+        if (ecdsaSig == NULL)
+            return WOLFSSL_FAILURE;
+        *siglen = wolfSSL_i2d_ECDSA_SIG(ecdsaSig, &sig);
+        wolfSSL_ECDSA_SIG_free(ecdsaSig);
+        return WOLFSSL_SUCCESS;
+    }
+#endif /* HAVE_ECC */
```

Comment:
If I understand your comment correctly, we can not check it in ```wolfSSL_EVP_PKEY_sign()```. Because ```wolfSSL_EVP_PKEY_sign()``` can not know the exact size of ```sig``` in it.

Reply:
`*siglen` should be set to the size of the `sig` buffer. Look at the `M_check_autoarg` check in OpenSSL. It gets called in `EVP_PKEY_sign` and does this exact check.

Reply:
> *siglen should be set to the size of the sig

Ok, fixed.

But there is an exceptional case. ```EVP_PKEY_sign()``` could be called to get ```siglen```. Then ```sig == NULL``` and ```*siglen``` is not checked.

This is a sample code from OpenSSL documentation.
https://www.openssl.org/docs/man1.1.1/man3/EVP_PKEY_sign.html

```
 /* Determine buffer length */
 if (EVP_PKEY_sign(ctx, NULL, &siglen, md, mdlen) <= 0)
     /* Error */

 sig = OPENSSL_malloc(siglen);

 if (!sig)
     /* malloc failure */

 if (EVP_PKEY_sign(ctx, sig, &siglen, md, mdlen) <= 0)
     /* Error */
```

Unfortunately existing wolfSSL ```EVP_PKEY_sign()``` does not set ```siglen``` when ```sig == NULL```, so I just add TODOs for now.

Reply:
Thanks for adding this. I implement size calculations and pushed a commit to your pull request. Please pull my commit before doing future work.